Source/DafnyLanguageServer/Workspace/DocumentDatabase.cs

```diff
@@ -135,29 +135,33 @@ ISymbolTableRelocator symbolTableRelocator
       }
     }
 
-    public Task<DafnyDocument> SaveDocumentAsync(TextDocumentIdentifier documentId) {
+    public async Task<DafnyDocument> SaveDocumentAsync(TextDocumentIdentifier documentId) {
       if (!documents.TryGetValue(documentId.Uri, out var databaseEntry)) {
-        return Task.FromException<DafnyDocument>(new ArgumentException($"the document {documentId.Uri} was not loaded before"));
+        throw new ArgumentException($"the document {documentId.Uri} was not loaded before");
       }
       if (!VerifyOnSave) {
-        return databaseEntry.Document;
+        return await databaseEntry.Document;
       }
-      return VerifyDocumentIfRequiredAsync(databaseEntry);
+      return await VerifyDocumentIfRequiredAsync(databaseEntry);
     }
 
     private async Task<DafnyDocument> VerifyDocumentIfRequiredAsync(DocumentEntry databaseEntry) {
-      var document = await databaseEntry.Document;
-      if (!RequiresOnSaveVerification(document)) {
-        return document;
+      try {
+        var document = await databaseEntry.Document;
+        if (!RequiresOnSaveVerification(document)) {
+          return document;
+        }
+        var cancellationSource = new CancellationTokenSource();
+        var updatedEntry = new DocumentEntry(
+          document.Version,
+          Task.Run(() => documentLoader.VerifyAsync(document, cancellationSource.Token)),
+          cancellationSource
+        );
+        documents[document.Uri] = updatedEntry;
+        return await updatedEntry.Document;
```

Comment:
Can you put the `try/catch` around  `documentLoader.VerifyAsync(document, cancellationSource.Token)` ?

Reply:
I'm not sure that the behavior in the `catch` block should be the same for all calls to `VerifyAsync`, the way things are currently structured. It's possible that we could refactor things so that it would be the same, though.

Reply:
It seems right now, if you cancel verification during `ApplyChangesAsync`, then it'll think the entire document load was cancelled and migrate the old document, even though only the verification was cancelled.

I think the behavior in the catch block should be the same for all calls to `VerifyAsync`, and not doing so may cause bugs.

Reply:
That does make a lot of sense to me. However, if I understand correctly, this PR preserves the existing behavior other than keeping verification from blocking. I think I'd prefer to make those changes as part of a separate PR which may be a pure refactor or may fix existing bugs other than the one this PR aims to fix.

Reply:
I think if you make the change I suggested, then you won't need the change in `TextDocumentLoader.Run`.

I think you could also only have the change in `TextDocumentLoader.Run`, without this change, but I believe @camrein didn't like that because it moves handling some cancellation from `DocumentDatabase` to `TextDocumentLoader`.

Reply:
Ah, but `TextDocumentLoader.Run` calls `VerifyInternal`, not `VerifyAsync`. That is, it does the actual verification work, rather than queuing up a task to run. And, from what I can tell, both the queuing and the execution need to be able to handle cancellation.

I originally tried modifying only `TextDocumentLoader.Run`, but it only halfway solved the problem.

Reply:
I think both of your changes handle cancellation that occurs during actual verification work. When the call to `VerifyAsync ` returns a Task, that task has already been queued. When you `await` that `Task`, a cancelled exception is thrown only when a cancelled exception is thrown inside `VerifyInternal` and there is caught and used in `request.Document.SetException(e);`.

Reply:
Ah, of course you're right. I've refactored with that in mind, according to your suggestions. I think this is the cleanest version of the fix so far, and it still seems to behave well.

Reply:
After closer investigation, it actually sometimes still behaves badly, but in a different way. Under some circumstances, it can be possible to get the IDE+server into a state where it responds with "Verification Succeeded" to essentially any change, including obvious syntax errors. This seems to occur after one or more verification runs have been cancelled before one finally succeeds.

My understanding was that returning the original document in the case of an `OperationCanceledException` was the right thing to do, but perhaps it could lead to the behavior I'm now seeing?

One additional data point is the closing and re-opening the file in VSCode seems to reset the state and let things behave correctly for a while.

Reply:
> After closer investigation, it actually sometimes still behaves badly, but in a different way. Under some circumstances, it can be possible to get the IDE+server into a state where it responds with "Verification Succeeded" to essentially any change, including obvious syntax errors. This seems to occur after one or more verification runs have been cancelled before one finally succeeds.

Do you know how to let VSCode show the LSP logs? Might be useful to get them in this case. @MikaelMayer wrote a nice help here: https://github.com/dafny-lang/ide-vscode/pull/92#issuecomment-996825783

Reply:
> But regardless, I do think your current change has improved things, so maybe let that second bug for another PR.

If you could share how I can reproduce this issue, I might better understand what's happening. More specifically, why my assumption about the root cause didn't resolve the issue. But I guess it's an issue that occasionally happens during development.

> I think one problem is that there is a state in the client, the contents of the status bar, that is not stored on the server.

This might be a possible reason. However, it's rather strange to me since the status notifications are sent with each change to the document. Have you guys checked the language server logs when this issue occurs? I'm just wondering if there's an exception thrown, such as this one:
https://github.com/dafny-lang/dafny/blob/189f4aef9ee49c80941d0cb9c3d20709a03a08da/Source/DafnyLanguageServer/Workspace/DocumentDatabase.cs#L97-L99
I was thinking if we ran in a race condition somewhere. However, all the queued `Task` instances that represent a change wait for their predecessor to complete (or handle its cancellation). Moreover, outside the `DocumentDatabase` changes cannot run in parallel since `TextDocumentLoader` just has a single thread to handle them. On the client side, VSCode is mostly an asynchronous event queue receiving the status notifications.

> In `HandleUpdateAndPublishDiagnosticsAsync` we can then send a notification for the current status bar state, both before and after `await documentAction`, and we won't need to send any notification in inner code like `VerifyInternal`.

If I understand you correctly, this will no longer allow showing intermediate states and will affect #1655. Or do want to send this information redundantly?

> I also think we should store a separate `Task<Document> ResolvedDocument` and `Task<Document> VerifiedDocument` in `DocumentEntry`, so that handlers for things like `go to definition` can execute while a document is still resolving.

Great idea +1
We could also experiment with running/awaiting the verification on a different thread to not block the document change queue.

Reply:
> If I understand you correctly, this will no longer allow showing intermediate states and will affect #1655. Or do you want to send this information redundantly?

The idea would be something like this:

```csharp
// Handles an LSP request
async Task ProcessRequest(Request request) {
  // One update for parsing+resolution, one for verification
  Func<State, Task<State>> stateUpdates = SplitIntoUpdates(request); 
  foreach(var stateUpdate in stateUpdates) {
    UpdateState(await stateUpdate(state));
  }
}

// Can also be called from internal state updates such as in #1655
private void UpdateState(State newState)
{
  if (!newState.StatusBar.Equals(state.StatusBar)) {
    publisher.SendStatusNotification(document, newState.StatusBar.Status, newState.StatusBar.Message);
  }
  state = newState;
}
```

Reply:
Ahh I see. You're suggesting that each step in the pipeline can be treated as a separate task within the request handlers. I like the approach of moving the LSP-related stuff more to the endpoints. But I think we should leave this architectural change to another PR (and maybe post Dafny 3.4.0).

Reply:
Sounds good :)